File: src/main/java/com/google/showcase/v1beta1/interview/q1/LinkedList.java
package com.google.showcase.v1beta1.interview.q1;

import java.util.function.Consumer;

interface LinkedList<T> {
    LinkedListNode<T> head();

    int size();

    void foreach(Consumer<T> block);

    void addLast(T value);

    void addFirst(T value);

    T removeLast();

    T removeFirst();
}

File: src/main/kotlin/com/google/showcase/v1beta1/interview/LinkedList.kt
package com.google.showcase.v1beta1.interview

import java.util.function.Consumer

interface LinkedListNode<T> {
    val next: LinkedListNode<T>?

    val prev: LinkedListNode<T>?

    val value: T
}

interface LinkedList<T> {
    val root: LinkedListNode<T>?

    fun size(): Int

    fun foreach(block: Consumer<T>)

    fun addLast(value: T)

    fun addFirst(value: T)

    fun removeLast(): T?

    fun removeFirst(): T?
}